### DISCUSSION TITLE
Not rendering everything in __all__ in init.py

### DISCUSSION THREAD

**Discussion by LiamConnors** [mkdocstrings/mkdocstrings#720] [Q&A] [answered]
Hi there! I'm currently exploring using mkdocstrings to build API reference docs from https://github.com/plotly/plotly.py/tree/master/packages/python/plotly/plotly/express and https://github.com/plotly/plotly.py/tree/master/packages/python/plotly/plotly/figure_factory

Adding the following to a markdown file doesn't render anything

```
::: packages.python.plotly.plotly.figure_factory
```

But the following does. It renders everything in `__all__`:

```
::: packages.python.plotly.plotly.express
```

Is there a reason in this case why it would pick up one but not the other?

**Comment by pawamoy** [comment]
Hey @LiamConnors! Cool to see Plotly devs experimenting with mkdocstrings ^^

I can't say right now what's going wrong, I'll investigate as soon as possible! Any specific branch I should checkout or master will do?

**Reply by LiamConnors** [reply]
Thanks so much @pawamoy! I've added the changes I've made to a branch here https://github.com/plotly/plotly.py/tree/api-docs The requirements to run the mkdocs site are defined in https://github.com/plotly/plotly.py/blob/api-docs/docs-requirements.txt if that's helpful.

**Comment by gvwilson** [comment]
thanks from me as well @pawamoy - it's a great tool and we're glad to have it.

**Comment by pawamoy** [comment]
There's a first thing that won't be well supported by mkdocstrings/Griffe: you inject documentation using `::: packages.python.plotly.etc`. I *think* you do this because you document things that aren't part of the actual package in `packages/python/plotly/plotly`. It might work but you'll get wrong identifiers (starting with `packages.python.plotly.plotly` instead of just `plotly`), meaning Plotly users won't be able to enjoy automatic or even manual cross-references to Plotly docs (intersphinx-like feature), since they import objects as `plotly.etc`, not `packages.python.plotly.plotly.etc`.

I tried changing that by removing the `packages.python.plotly.` prefix from `:::` instruction, but then you try to inject things that aren't part of the package, like `_plotly_utils.colors`. I wonder: are there utilities part of the final package? If so, how do users import them? Maybe they don't, and this layout is hidden to them? I guess `plotly` itself does some `from _plotly_utils import ...`?

**Comment by pawamoy** [comment]
With this patch:

````diff
diff --git a/docs/reference/graph_objects/Figure.md b/docs/reference/graph_objects/Figure.md
index 757267a68..49c877415 100644
--- a/docs/reference/graph_objects/Figure.md
+++ b/docs/reference/graph_objects/Figure.md
@@ -5,7 +5,7 @@ import plotly.graph_objects as go
 fig = go.Figure()
 ```
 
-::: packages.python.plotly.plotly.graph_objs
+::: plotly.graph_objs
     options:
         merge_init_into_class: true
         inherited_members: true
diff --git a/docs/reference/graph_objects/Layout.md b/docs/reference/graph_objects/Layout.md
index 5d4873e59..a5db4b37f 100644
--- a/docs/reference/graph_objects/Layout.md
+++ b/docs/reference/graph_objects/Layout.md
@@ -1,3 +1,3 @@
-::: packages.python.plotly.plotly.graph_objs.Layout
+::: plotly.graph_objs.Layout
     options:
         merge_init_into_class: true
diff --git a/docs/reference/graph_objects/traces/scatter.md b/docs/reference/graph_objects/traces/scatter.md
index 877a1d5d5..9b5d524e6 100644
--- a/docs/reference/graph_objects/traces/scatter.md
+++ b/docs/reference/graph_objects/traces/scatter.md
@@ -1,4 +1,4 @@
 # graph_objects.Scatter
 
-::: packages.python.plotly.plotly.graph_objs.Scatter
+::: plotly.graph_objs.Scatter
 
diff --git a/docs/reference/plotly_colors.md b/docs/reference/plotly_colors.md
index dc9bdd894..4727181f4 100644
--- a/docs/reference/plotly_colors.md
+++ b/docs/reference/plotly_colors.md
@@ -1 +1 @@
-::: packages.python.plotly._plotly_utils.colors
+::: _plotly_utils.colors
diff --git a/docs/reference/plotly_data.md b/docs/reference/plotly_data.md
index 23ff765c2..af5a4f2fb 100644
--- a/docs/reference/plotly_data.md
+++ b/docs/reference/plotly_data.md
@@ -1,2 +1,2 @@
-::: packages.python.plotly.plotly.data
+::: plotly.data
 
diff --git a/docs/reference/plotly_express/index.md b/docs/reference/plotly_express/index.md
index 59b9912da..5269a6798 100644
--- a/docs/reference/plotly_express/index.md
+++ b/docs/reference/plotly_express/index.md
@@ -3,5 +3,5 @@
 The `plotly.express <https://plot.ly/python/plotly-express/>` module is
 plotly's high-level API for rapid figure generation.
 
-::: packages.python.plotly.plotly.express
+::: plotly.express
 
diff --git a/docs/reference/plotly_express/subpackages/colors.md b/docs/reference/plotly_express/subpackages/colors.md
index ad33155b4..fb99ac6f7 100644
--- a/docs/reference/plotly_express/subpackages/colors.md
+++ b/docs/reference/plotly_express/subpackages/colors.md
@@ -1,3 +1,3 @@
-<!-- ::: packages.python.plotly._plotly_utils.colors -->
+<!-- ::: _plotly_utils.colors -->
 
-::: packages.python.plotly.plotly.express.colors
\ No newline at end of file
+::: plotly.express.colors
\ No newline at end of file
diff --git a/docs/reference/plotly_express/subpackages/data.md b/docs/reference/plotly_express/subpackages/data.md
index 78ac9e7c7..a8b4c8685 100644
--- a/docs/reference/plotly_express/subpackages/data.md
+++ b/docs/reference/plotly_express/subpackages/data.md
@@ -1,4 +1,4 @@
 # plotly.express.data
 
-::: packages.python.plotly.plotly.data
+::: plotly.data
 
diff --git a/docs/reference/plotly_express/subpackages/imshow_utils.md b/docs/reference/plotly_express/subpackages/imshow_utils.md
index 32c1f0ffe..ec2f5287c 100644
--- a/docs/reference/plotly_express/subpackages/imshow_utils.md
+++ b/docs/reference/plotly_express/subpackages/imshow_utils.md
@@ -1,2 +1,2 @@
-::: packages.python.plotly.plotly.express.imshow_utils
+::: plotly.express.imshow_utils
 
diff --git a/docs/reference/plotly_express/subpackages/trendline_functions.md b/docs/reference/plotly_express/subpackages/trendline_functions.md
index ce8efef4d..80cf2316b 100644
--- a/docs/reference/plotly_express/subpackages/trendline_functions.md
+++ b/docs/reference/plotly_express/subpackages/trendline_functions.md
@@ -1,4 +1,4 @@
 # plotly.express.trendline_functions
 
-::: packages.python.plotly.plotly.express.trendline_functions
+::: plotly.express.trendline_functions
 
diff --git a/docs/reference/plotly_figure_factory.md b/docs/reference/plotly_figure_factory.md
index 46cd6417a..31577af6e 100644
--- a/docs/reference/plotly_figure_factory.md
+++ b/docs/reference/plotly_figure_factory.md
@@ -1,2 +1,2 @@
-::: packages.python.plotly.plotly.figure_factory
+::: plotly.figure_factory
 
diff --git a/docs/reference/plotly_io.md b/docs/reference/plotly_io.md
index 69ebb214f..b1007ec0d 100644
--- a/docs/reference/plotly_io.md
+++ b/docs/reference/plotly_io.md
@@ -1 +1 @@
-::: packages.python.plotly.plotly.io
+::: plotly.io
diff --git a/docs/reference/plotly_subplots.md b/docs/reference/plotly_subplots.md
index 3ad521e72..49d70ae28 100644
--- a/docs/reference/plotly_subplots.md
+++ b/docs/reference/plotly_subplots.md
@@ -1,3 +1,3 @@
-::: packages.python.plotly.plotly.subplots
+::: plotly.subplots
 
 
diff --git a/mkdocs.yml b/mkdocs.yml
index f12ea8a01..9808cd898 100644
--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -31,6 +31,8 @@ plugins:
     default_handler: python
     handlers:
       python:
+        paths:
+        - packages/python/plotly
         options:
           heading_level: 3
           show_source: false
````

...we get this rendered page:

![figure_factory](https://github.com/user-attachments/assets/0028aed3-48d3-419e-9dc1-628e2abd3208)

Does it look better?

**Reply by LiamConnors** [reply]
Thanks @pawamoy! Yes, that shows what I'd expect to see. I hadn't realized there was a difference between injecting it using `packages/python/plotly/plotly` vs `plotly` and then setting `paths`. 

Using `::: packages.python.plotly._plotly_utils.colors` was only to see if it would bring in more information about the `colors` package, which a user could access with `from plotly import colors`. 

Using `::: packages.python.plotly.plotly.colors` (or now `::: plotly.colors`) only brought in the description at the top of 
https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/plotly/colors/__init__.py but not the functions like you'd see in Python help:

```
>>> from plotly import colors
>>> help(colors)
```

Those functions are currently defined in https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/_plotly_utils/colors/__init__.py and then imported in https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/plotly/colors/__init__.py 
Looking further at this I see that's `_plotly_utils` is another package that is included here in `setup.py` https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/setup.py#L550). I guess that setup may be the reason it doesn't work.

**Reply by pawamoy** [reply, marked as answer]
> I hadn't realized there was a difference between injecting it using packages/python/plotly/plotly vs plotly and then setting paths.

Here are the docs https://mkdocstrings.github.io/python/usage/#finding-modules :slightly_smiling_face: 

> Using ::: packages.python.plotly.plotly.colors (or now ::: plotly.colors) only brought in the description at the top of
https://github.com/plotly/plotly.py/blob/master/packages/python/plotly/plotly/colors/__init__.py but not the functions like you'd see in Python help:

The technical explanation is: `plotly.colors` imports everything from `_plotly_utils.colors`, and `_plotly_utils` is another package, so Griffe isn't able to resolve the aliases in `plotly.colors` (the wildcard imported names). To fix this, you can [preload](https://mkdocstrings.github.io/python/usage/configuration/general/?h=preload#preload_modules) `_plotly_utils`:

```yaml
plugins:
- mkdocstrings:
    default_handler: python
    handlers:
      python:
        paths:
        - packages/python/plotly
        options:
          heading_level: 3
          show_source: false
          show_bases: false
          show_symbol_type_heading: true
          docstring_style: numpy
          merge_init_into_class: true
          preload_modules:
          - _plotly_utils
          - _plotly_future_  # maybe it's needed too?
```

With this you can replace `::: _plotly_utils.colors` with `::: plotly.colors` :slightly_smiling_face: 

The alternative "brute-force" solution is to set [`load_external_modules: true`](https://mkdocstrings.github.io/python/usage/#load_external_modules) (which I think you tried to use, but it's a global-only option and is therefore expected at the same level as `paths`, not under `options`), but this isn't recommended, as it quickly results in the entire standard library being loaded (to resolve names), which isn't super efficient :smile:

**Reply by LiamConnors** [reply]
Thanks. Sorry, missed that part in the docs that explained  for "package in subfolder". This makes sense with preloading those modules. And works perfectly.

**Reply by pawamoy** [reply]
No worries! Glad I could help :slightly_smiling_face:

**Comment by pawamoy** [comment]
While I'm at it: I see a few warnings in MkDocs' output like:

```
WARNING -  mkdocs_autorefs: reference/plotly_subplots.md: from /media/data/dev/plotly.py/packages/python/plotly/plotly/subplots.py:27: (plotly.subplots.make_subplots) Could not find cross-reference target ' (2,1) xaxis2,yaxis2 '
```

This is because you're using the `>>>` pycon syntax without fences (triple backticks), so Python-Markdown interprets it as markup, and mkdocs-autorefs thinks that things like `[i][1]` are cross-refs.

To fix this you could use this extension: https://pawamoy.github.io/markdown-pycon/, which allows such pycon syntax outside of code blocks :slightly_smiling_face: 

There are other false-positive like:

````md
    This is the format of your plot grid:
    [ (1,1) xaxis1,yaxis1 ]
    [ (2,1) xaxis2,yaxis2 ]
````

...which can be fixed by escaping that as code blocks:

````md
    This is the format of your plot grid:

    ```
    [ (1,1) xaxis1,yaxis1 ]
    [ (2,1) xaxis2,yaxis2 ]
    ```
````

**Reply by pawamoy** [reply]
Hmmm I notice a priority issue where things like `i[0][1]` in non-fenced pycon blocks are still handled by autorefs, when they shouldn't. I've opened https://github.com/pawamoy/markdown-pycon/issues/1.
